### ISSUE TITLE
feature request: rendering blocks/partials as strings to pipe into other functions

### ISSUE THREAD

**Issue by prydonius** [kubernetes/helm#1005]
Currently, if you want to add a partial you need to create a helper and define templates:
For example:
```
{{ define "nginx.vhost" }}
    server {
      location / {
        proxy_pass http://{{ .Values.address }}
      }
    }
{{ end }}
```

and then used in a manifest like this:
```
...
data:
  vhost: |
    {{ template "nginx.vhost" . }}
```

This issue is that I have to manually indent in the define block, because template doesn't return a string to pipe to `indent`.

A much easier to use partial system would be to use the files map from #994, so I can create
vhost.conf:
```
server {
  location / {
     proxy_pass http://{{ .Values.address }}
  }
}
```

and then have:

```
...
data:
  vhost: |
    {{ .Files.Get "files/vhost.conf" | indent 4 }}
```

If we don't want files to be used as partials (it would be strange to have two different ways of using partials), is there a way the output from `template` can be piped to other functions?

**Comment by technosophos** [comment]
So there are some really weird edge cases where the template partial has to be compiled before the main template function is executed. (The edge cases mostly involve not being able to access other templates, so things like `block`, `define`, and `template` can produce unexpected results).  So just using `Files.Get` won't be possible. But there might be another way.

Right now, I'm trying to figure out if I can just add a template function to the func map that will execute one of the pre-compiled template partials, capture the output, and return it as text. That would effectively give you what you're asking for. So it would look something like this:

```
{{ partial "nginx.vhost" . | indent 4 }} 
```

I'm not _totally_ sure I can make this work, but I'm going to give it a try.

**Comment by migmartri** [comment]
Hi,

I am not in favor of creating a new function to enhance the `template` one, taking into account that:

* The problem that it solves is solvable (painfully at least) adding spaces in the template.
* It adds another function that can confuse people, when to use `partial` and when `template`? 
* There is already another term called "partial templates" (.tpl) files.

That said, I think that the proposed functionality is useful in the long run, so the way I think of the partial function is as   **a replacement** of the template function in which case:

* People will use it as the only way to render partial templates and it will be perceived as a template function on steroids.
* This function allows other piped transformations.
* It is decoupled (at least in the signature) from Go templates.
* The linter could alert the user of the usage of this new function instead. 

In any case, for the current Helm version, I am not sure if it makes sense to make these kind of changes in the template engine.

At least this can act as a base for future discussion on what's missing in the current templating engine.

Cheers

**Comment by technosophos** [comment]
@migmartri how about if I rename the function to `include`?

Then it would work like this:

```
{{include "conrad/_partial" . | indent 2}}
``` 

I agree with the suggestion that we encourage people to use `include` instead of `template` so that we can control output. Unfortunately, since `template` is a builtin and not a function, there's nothing we can do to mask it. So Linter's our best bet.

**Comment by migmartri** [comment]
> {{include "conrad/_partial" . | indent 2}}

Makes sense to me :)

> Unfortunately, since template is a builtin and not a function, there's nothing we can do to mask it. So Linter's our best bet.

A linter warning will be enough to let people know that there is a `helm` way to do it :)

**Comment by technosophos** [comment]
Alright. I made the changes to the PR. Really, it's up to ya'll whether you want me to commit the change. On the plus side, it will make things like indenting a lot easier. But the negative is that it could be confusing to people.

**Comment by migmartri** [comment]
TLDR; Ship it! :)

If we all agree that `include` is the way to go, the sooner we land the code and start using it the better since we do not have a set of "final" charts yet and they can be changed.

Changing the linter to raise a message should not be difficult either but in my opinion we should wait until we are certain that this function can replace `template` in stability and performance as well.

**Comment by prydonius** [comment]
:+1: from me, and I agree with @migmartri's comments about the linter.

**Comment by technosophos** [comment]
Okay, cool. Thanks!
